### PR TITLE
[hmac,dv] FIFO empty interrupt and test mode pins

### PR DESCRIPTION
**Pull request by martin-velay**
- Verify FIFO empty interrupt register to tackle issue #21815
  - Add checks on interrupt pins when test mode is triggered via the dedicated register INTR_TEST

### CODE REVIEW

**Comment by martin-velay** [comment]
I still need to run a full regression to be sure nothing is broken, but you could start to review it

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv [hunk 1]
```diff
@@ -193,14 +206,16 @@ class hmac_scoreboard extends cip_base_scoreboard #(.CFG_T (hmac_env_cfg),
             end
           end
           "intr_test": begin // testmode, intr_state is W1C, cannot use UVM_PREDICT_WRITE
-            bit [TL_DW-1:0] intr_state_exp = item.a_data | `gmv(ral.intr_state);
+            intr_test = item.a_data;
+            // As hmac_err and hmac_done are W1C, we should get back the current state
```
Comment (rswarbrick) [inline hw/ip/hmac/dv/env/hmac_scoreboard.sv:210]: I think the comment might be talking about the wrong registers here?

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv [hunk 2]
```diff
@@ -537,21 +554,96 @@ class hmac_scoreboard extends cip_base_scoreboard #(.CFG_T (hmac_env_cfg),
...
+
+      // Delay FIFO empty signal to be aligned with the DUT behavior
+      fork
+        begin
+          cfg.clk_rst_vif.wait_clks(1);
+          fifo_empty_intr = pre_fifo_empty_intr;
         end
+      join_none
+
+      // Check wether FIFO full has been detected for the ongoing message but the retrictions cases
```
Comment (rswarbrick) [inline hw/ip/hmac/dv/env/hmac_scoreboard.sv:605]: Nit: Spelling of whether

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv [hunk 3]
```diff
@@ -537,21 +554,96 @@ class hmac_scoreboard extends cip_base_scoreboard #(.CFG_T (hmac_env_cfg),
...
+
+      // Compute FIFO level flags
+      hmac_fifo_depth = hmac_wr_cnt - hmac_rd_cnt;
+      hmac_fifo_full  = hmac_fifo_depth == HMAC_MSG_FIFO_DEPTH_WR;
+      hmac_fifo_empty = hmac_fifo_depth == 0;
+
+      // The FIFO empty interrupt is raised only if the message FIFO is actually writable by
+      // software, i.e., if all of the following conditions are met:
+      //   1- The HMAC block is not running in HMAC mode and performing the second round of
+      //      computing the final hash of the outer key as well as the result of the first round
```
Comment (rswarbrick) [inline hw/ip/hmac/dv/env/hmac_scoreboard.sv:584]: I'm not an HMAC expert but this looks surprising to me. Do you want to drop the "as well" here?

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv [hunk 4]
```diff
@@ -537,21 +554,96 @@ class hmac_scoreboard extends cip_base_scoreboard #(.CFG_T (hmac_env_cfg),
...
+        pre_fifo_empty_intr = 0;
+      end
+
+      // Delay FIFO empty signal to be aligned with the DUT behavior
+      fork
+        begin
+          cfg.clk_rst_vif.wait_clks(1);
+          fifo_empty_intr = pre_fifo_empty_intr;
         end
+      join_none
```
Comment (rswarbrick) [inline hw/ip/hmac/dv/env/hmac_scoreboard.sv:603]: I can imagine this getting rather confusing if something surprising happens in the middle, like a reset stopping the clock for a bit. Would it make more sense to do this explicitly and copy `pre_fifo_empty_intr` to `fifo_empty_intr` at the start of this task? (since this task is the only one that currently uses it)

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv [hunk 5]
```diff
@@ -537,21 +554,96 @@ class hmac_scoreboard extends cip_base_scoreboard #(.CFG_T (hmac_env_cfg),
...
+  endtask : hmac_process_fifo_status
+
+  // Check interrupt pins in test mode
+  virtual task hmac_intr_test();
+    bit [TL_DW-1:0] intr_en;
+    hmac_intr_e     intr;
+    forever @(intr_test) begin
+      intr_en = `gmv(ral.intr_enable);
+      foreach (intr_test[intr_i]) begin
+        fork begin
```
Comment (rswarbrick) [inline hw/ip/hmac/dv/env/hmac_scoreboard.sv:623]: The fork structure here is a bit hard to follow. I think it might be easier if you split this task into "one per bit", so `intr_i` ends up as an argument to the task. Then the chopped up task can wait with `@(intr_test[intr_i])` and all the fork stuff (and things like having to explicitly chop out `auto_intr_i`) ends up outside of that task.

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv (hunk 1 above)
Comment (martin-velay) [reply hw/ip/hmac/dv/env/hmac_scoreboard.sv:210]: Here I am referring to the INTR_STATE register as it is predicted while we are accessing the INTR_TEST register.
I have changed a bit my comment to make it clearer.

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv (hunk 2 above)
Comment (martin-velay) [reply hw/ip/hmac/dv/env/hmac_scoreboard.sv:605]: Sorry, that one doesn't want to stick in my mind!

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv (hunk 3 above)
Comment (martin-velay) [reply hw/ip/hmac/dv/env/hmac_scoreboard.sv:584]: This comment is almost a copy/paste of the description here:
https://opentitan.org/book/hw/ip/hmac/doc/registers.html#intr_state--fifo_empty

This description is from @vogelpi. What do you think?

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv (hunk 4 above)
Comment (martin-velay) [reply hw/ip/hmac/dv/env/hmac_scoreboard.sv:603]: No, I can't do this copy at the beginning of this task it require the updated `pre_fifo_empty_int` value at this exact moment.
But I can do like the other tasks and implement a fork like:

```
            begin : reset_hmac
              wait(cfg.under_reset);
            end
          join_any
          disable fork;
```

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv (hunk 5 above)
Comment (martin-velay) [reply hw/ip/hmac/dv/env/hmac_scoreboard.sv:623]: OK, it's done. I hope it'll be easier to read and also that I haven't broken anything.

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv (hunk 4 above)
Comment (martin-velay) [reply hw/ip/hmac/dv/env/hmac_scoreboard.sv]: I have pushed a commit which doesn't work yet! https://github.com/lowRISC/opentitan/pull/23749/commits/dbf5ec3c203d3f506c65af01c9fd0fe21fd8778e

I'll work on it later next week.

File: hw/ip/hmac/dv/env/hmac_scoreboard.sv (hunk 3 above)
Comment (vogelpi) [reply hw/ip/hmac/dv/env/hmac_scoreboard.sv:594]: Sorry, my description is probably confusing. What I meant here is:
If HMAC is running in HMAC mode (vs. the unkeyed SHA2 mode) and currently doing the second round, software cannot write the FIFO. During this second round, HMAC hashes the outer key together with the result of the first round. The first round result was produced using the inner key.

**Comment by martin-velay** [comment]
@vogelpi I have added a temporary workaround to be able to merge without taking the related RTL fix from this PR https://github.com/lowRISC/opentitan/pull/23739.

Now, this PR could be merged into the master and won't break anything.

**Comment by martin-velay** [comment]
@vogelpi, as it has been decided to merge the RTL fix from this PR #23739 , I have reverted the proposed workaround mentioned above.
If needed, it could be reestablished, see this commit: https://github.com/lowRISC/opentitan/commit/55e20274ed386a1ad7ff6aa453ea9440ff39f49b

**Comment by rswarbrick** [comment]
Hi @martin-velay! Would you mind rebasing this over recently merged changes?

**Review by vogelpi — APPROVED** [review]
LGTM, when rebasing, would mind correcting the commit messages? One commit contains two sign-off lines.

**Comment by rswarbrick** [comment]
Sorry, I jumped the gun! It turns out that the rebasing wasn't too hard. Happy for either of us to merge, but maybe it makes sense to wait long enough to check that I didn't accidentally break the compile.

**Comment by vogelpi** [comment]
Thanks @rswarbrick . The DV checks passed already. What's left are the FPGA tests which are not affected by this. I am thus merging the PR now.
